File: 892.Surface-Area-of-3D-Shapes.py
# https://leetcode.com/problems/surface-area-of-3d-shapes/description/
#
# algorithms
# Esay (56.2%)
# Total Accepted:    1.6k
# Total Submissions: 2.9k


class Solution(object):
    def surfaceArea(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        N = len(grid)
        if N == 1:
            return 0 if grid[0][0] == 0 else 2 + grid[0][0] * 4

        res = 0

        for i in xrange(N):
            for j in xrange(N):
                if grid[i][j] != 0:
                    res += 2
                if i == 0 or i == N - 1:
                    res += grid[i][j]
                if j == 0 or j == N - 1:
                    res += grid[i][j]
                if i > 0:
                    res += abs(grid[i][j] - grid[i - 1][j])
                if j > 0:
                    res += abs(grid[i][j] - grid[i][j - 1])

        return int(res)

File: 892.Surface-Area-of-3D-Shapes.java
// https://leetcode.com/problems/surface-area-of-3d-shapes/
//
// algorithms
// Easy (57.73%)
// Total Accepted: 15,255
// Total Submissions: 26,425


class Solution {

    private static final int[][] deltaArr = new int[][] { { -1, 0 }, { 1, 0 }, { 0, -1 }, { 0, 1 } };

    public int surfaceArea(int[][] grid) {
        int res = 0;
        int row = grid.length, col = grid[0].length;

        for (int x = 0; x < row; x++) {
            for (int y = 0; y < col; y++) {
                int num = grid[x][y];
                if (num == 0) {
                    continue;
                }

                res += 2;

                int diff;
                for (int[] delta : deltaArr) {
                    if (isValidCoordinate(row, col, x + delta[0], y + delta[1])) {
                        diff = num - grid[x + delta[0]][y + delta[1]];
                        if (diff > 0) {
                            res += diff;
                        }
                    } else {
                        res += num;
                    }
                }
            }
        }

        return res;
    }

    public boolean isValidCoordinate(int row, int col, int x, int y) {
        if (x < 0 || x >= row) {
            return false;
        }

        if (y < 0 || y >= col) {
            return false;
        }

        return true;
    }
}